icd/api/pipeline_compiler.cpp

```diff
@@ -697,6 +790,53 @@ VkResult PipelineCompiler::CreateGraphicsPipelineBinary(
 #endif
     }
 
+#if XGL_USE_EXPERIMENTAL_SHADER_CACHE_PIPELINES
+    PipelineCompilerType cacheType = GetShaderCacheType();
+    Util::Result cacheResult = Util::Result::Success;
+    if (pPipelineCache != nullptr)
+    {
+        Llpc::IShaderCache* const pLlpcShaderCache =
+            pPipelineCache->GetShaderCache(deviceIdx).GetCachePtr().pLlpcShaderCache;
+
+        Util::MetroHash128 hash = {};
+
+        if (pLlpcShaderCache != nullptr)
+        {
+            hash.Update(pipelineHash);
+            hash.Update(pCreateInfo->pipelineInfo.vs.options);
+            hash.Update(pCreateInfo->pipelineInfo.tes.options);
+            hash.Update(pCreateInfo->pipelineInfo.tcs.options);
+            hash.Update(pCreateInfo->pipelineInfo.gs.options);
+            hash.Update(pCreateInfo->pipelineInfo.fs.options);
+            hash.Update(pCreateInfo->pipelineInfo.options);
+            hash.Update(pCreateInfo->flags);
+            hash.Update(pCreateInfo->dbFormat);
+            hash.Update(pCreateInfo->pipelineProfileKey);
+            hash.Update(deviceIdx);
+            hash.Update(pCreateInfo->compilerType);
+            hash.Finalize(pCacheId->bytes);
+        }
+
+        bool isHit = false;
+        if (pLlpcShaderCache != nullptr && cacheType == PipelineCompilerTypeLlpc)
+        {
+            cacheResult = RetrievePipeline(
+                pCacheId, pPipelineBinarySize, ppPipelineBinary, pInstance, pLlpcShaderCache);
+            m_cacheAttempts++;
+
+            isHit = (cacheResult == Util::Result::Success);
+        }
+        if (isHit)
+        {
+            pCreateInfo->elfWasCached = true;
+            shouldCompile             = false;
+            m_cacheHits++;
+            pCreateInfo->pipelineFeedback.hitApplicationCache = true;
+        }
+    }
+#endif // XGL_USE_EXPERIMENTAL_SHADER_CACHE_PIPELINES
+
+#if VKI_USE_PAL_PIPELINE_CACHE
```

Comment:
Can we make mutex between XGL_USE_EXPERIMENTAL_SHADER_CACHE_PIPELINES and VKI_USE_PAL_PIPELINE_CACHE?
Here should be \#else ?

Reply:
These two are mutually exclusive, i.e., VKI_USE_PAL_PIPELINE_CACHE is defined to true iff XGL_USE_EXPERIMENTAL_SHADER_CACHE_PIPELINES is set.

Do you suggest to remove the VKI_USE_PAL_PIPELINE_CACHE and have if else everywhere?

Reply:
Yes, it's my opinion.
We don't need introduce VKI_USE_PAL_PIPELINE_CACHE more， and tow cache are enabled at same time, the performance could be impacted.

Reply:
I'd rather keep this macro to stay minimize the amount of changes compared to the proprietary driver. VKI_USE_PAL_PIPELINE was introduced introduced in these sources 2 years ago by a merge by Ping Fu.

Let me know if I should remove it anyway or can keep it.

Reply:
VKI_USE_PAL_PIPELINE should already be removed when we open source pipeline binary cache in my memory.
I also grep it in whole amdvlk repo, I didn't see it.
Anything I still missed?

Reply:
Hmm, perhaps our sources diverged here. If that's the case, I'll remove it in favor of #if #else, like you recommended.